### ISSUE TITLE
elasticsearch.js call to index/create hangs at 277 documents

### ISSUE THREAD

**Issue by yipmhw** [elastic/elasticsearch-js#162]
Hi all,

I'm writing a web crawler in Node.js and indexing with ElasticSearch. However, I've ran into a problem where the code hangs at the indexing function. The call back function appears to have been executed but the function does not return fully and just blocks the code. I have also tried using the client.create API call and the same thing happens.

Here's how the client is initialised (just trial and error with settings):

``` js
var es_client = new elasticsearch.Client({
    host: "localhost:9200",
    log: ['error', 'trace'],
    keepAlive: true,
    sniffOnConnectionFault: true,
    //sniffInterval: 6000,
    sniffOnStart: true,
    maxKeepAliveTime: 600000
});
```

Here's the indexing API call that hangs:

``` js
es_client.index({
  index: seedURL,
  type: 'post',
  id: generate_md5(username + "\n" + post_title + "\n" + post_content),
  body: {
    thread_md5 : thread_md5,
    thread_title : thread_title,
    thread_url : post_list_page_url,
    post_title: post_title,
    post_order : post_order,
    post_content: post_content,
    timestamp: timestamp,
    username: username
  }
}).then(
    function (resp) {
        console.log("Elasticsearch response to indexing " + post_title + "...");
        console.log(resp);
    }, 
    function (err) {
        console.log("[ERROR] An error occurred whilst indexing: " + post_title + "...");
        console.log(err.message);
    }
);
```

For contextual purposes, it may be helpful to know that the index/create API call is called from within a http.get callback.

I have been testing the crawler script by commenting out the API call to indexing and it finishes the crawl no problem. This showed that the problem somehow lies with ElasticSearch JS.

I have also had a look at the ElasticSearch logs and no errors were raised.

Lastly - and this could be the best hint yet - is that the number of documents successfully indexed at every trial run hangs at exactly 277 documents.

Thoughts?

Many thanks for your help in advance.

**Comment by Hosar** [comment]
I was having the same issue, but only when the connection was refused.

Elasticsearch ERROR: 2015-04-05T01:08:07Z  Error: Request error, retrying -- connect ECONNREFUSED

I just needed to close the client. **client.close();**

    var simpleGet = function(client){	
	client.get({
	  index: 'megacorp',
	  type: 'employee',
	  id: '1'
	}, function (error, response) {
		if(error){
		  console.log(error);	 	
	 	  client.close();
		}
	});			
    }

**Comment by shadowcodex** [comment]
Is it possible that you could close the client before a call is complete and cut off the call in the middle of it?

example: if you called an API, then it returned a bunch of json objects. You iterate through each object to store into elastic search. After you fired off all of the creates you call client.close(). However what if all the creates are finished? Will this cause an issue?

The event driven model really messes with me haha.

**Comment by sean-liang** [comment]
I got similar issue with the the version v10.1.3. It makes the application hang after 2~3 minutes. I tried same logic with other storage options like leveldb, it works fine.

File:
https://github.com/sean-liang/p2pspider/tree/master/samples

Btw, I'm using below command to start a elasticsearch server:
`docker run --name es -d -p 9200:9200 -p 9300:9300  -v /home/sean/docker/elasticsearch/esdata:/usr/share/elasticsearch/data elasticsearch`
